🐛 amp-ima-video: Show sound and fullscreen buttons during ads

curseagain opened the pull request:
Closes https://github.com/ampproject/amphtml/issues/18533

## Changes

### First Commit

- Controls are always visible while ads are playing.
- The only buttons displayed in ad control panel are `#ima-mute-unmute` and `#ima-fullscreen`.
- Controls are restored after the ad finishes.

### Second Commit

- The ad countdown timer has been moved to the control panel, is displayed when the ad controls are displayed and hidden when the ad controls are hidden

### Third Commit

- The ad countdown timer text can be customized by setting `data-ad-label` attribute to a pattern (e.g. `data-ad-label="Publicidad (%s de %s)"`, resulting in something like "`Publicidad (1 de 2): 0:27`")
- Added additional amp-ima-video test for ad with skip button
- Note: The height of the ad label is fixed and any overflow is hidden

### Fourth Commit

- Display smaller controls when an ad is skippable and has mobile classes, signifying that the "Skip" button will be displayed and we will have to squish our controls underneath.
- Fix bug that caused ad to be `videoWidth` instead of `fullscreenWidth` when fullscreen is toggled

## Screenshots

### Desktop with Custom `data-ad-label`

<img src="https://user-images.githubusercontent.com/4807680/48569625-7a1e3e00-e8bf-11e8-82d3-94a3500b52bc.gif" height="300"/>

---

### Desktop with Skip Button

<img src="https://user-images.githubusercontent.com/4807680/48569641-84d8d300-e8bf-11e8-9bdd-a6a7061a7acf.gif" height="300"/>

---

### Mobile without Skip Button

<img src="https://user-images.githubusercontent.com/4807680/48729034-38153500-ebfc-11e8-86dd-4266a34cdd2f.gif" height="300"/>

---

### Mobile with Skip Button

<img src="https://user-images.githubusercontent.com/4807680/48728893-e5d41400-ebfb-11e8-8f09-48e41662e1e6.gif" height="300"/>

curseagain commented:
@torch2424 this is ready for review!  However, I am concerned that the `.videoAdUiAttribution` element within the ad iframe is not clearly visible underneath the controls. See the screenshot in this PR.  What do you think?  Is that an ad disclosure that we aren't allowed to cover?

torch2424 commented:
Wow this was fast! Awesome, thank you very much for this! 😄 Will look over this, and reach out to some people on our end.

I'll also ask around the for the `.videoAdUiAttribution` element for you 👍 

cc @jasti

torch2424 commented:
Also, did some quick research, was suggested by @aghassemi that following the [VideoJS IMA Plugin](https://github.com/googleads/videojs-ima) would be a great place to start.

Looking at [their simple example](https://googleads.github.io/videojs-ima/examples/simple/) It appears that their `.videoAdUiAttribution` is brought above the controls shade. So I think we should go ahead and do the same.

Tried looking around for more reference, but couldn't find much. Hopefully will know more later today 😄

curseagain commented:
@torch2424, it looks like the videojs team has created their own countdown timer to replace the default one, as described [here](https://developers.google.com/interactive-media-ads/docs/sdks/html5/countdown-timer).

That works fine, but with one complication: internationalization!  The videojs player accepts [custom values for `adLabel` and `adLabelNofN`](https://github.com/googleads/videojs-ima/blob/0fbf8f5321f85fba294f64c3f311333feb114f4e/README.md), which we could potentially do as well...  However, maybe there are similar translations happening elsewhere that we could refer to?

curseagain commented:
@torch2424 I've added a second commit in which I do some linting and attempt to add an ad countdown timer like they do in the videojs plugin.  I've also attached a screenshot and updated the description in the PR to reflect the current problem: translation of the ad disclosure text.

torch2424 commented:
@curseagain Thanks for all the videos! Really helpful!

Glanced over the code again, really like they way you implemented. Once we get finalized on an implementation I'll give a more thorough review (though I don't think I'll have too many comments). 😄 

So, the Ads person we work with, is currently Out of Office until next week I think, so I sent them and email and then they can help advise us on this. 

Thanks for making that countdown timer, and looking into that! Looks great!

i18n will definitely be a problem, and thank you very much for considering that. Once we hear back from our Ads contact, hopefully we can learn if there is something to do that for us. If not, we can discuss a strategy to move forward on i18n.

torch2424 commented:
@curseagain 

Just a quick update:

Got a response from the IMA team, but they sent me to ask someone else about the video controls and i18n haha! So now we will wait for that 😄

torch2424 commented:
@curseagain 

Got a super quick response from the IMA team 😄 

"I would definitely hide when not hover. Other than that it looks fine for the most part. I would also try to minimize the shadow size of the controlbar so as to not interfere with the skip button and things like that".

So, once we get those implemented, I think we can pull that in 😄 We can either do the "show controls on hover" in another PR, merge to master, pull into this PR, and then go from there. Or I wouldn't mind just adding it to this PR.

I'm kind of low on bandwidth this sprint, so if you have time definitely feel free to get started on this. If you cannot, perhaps sometime next week, I can open a PR for the controls hover 😄 

Also, just asked about i18n as well.

Thanks!

curseagain commented:
Thanks @torch2424!  I can begin on some of the other tweaks while we wait to know more about i18n.

torch2424 commented:
@curseagain 

So I brought up i18n at Today's design review #18658 (Thank you so much for bringing it up in the issue though, totally forgot that it was a good place to discuss there 🙃) 

They said we could make i18n an attribute on the component itself! They said for consistency's sake, we could/should follow what [amp-carousel](https://www.ampproject.org/docs/reference/components/amp-carousel) is doing with their `data-button-count-format`. 

We can simply pass in that attribute value into the ima video player like we are the other `data-` attributes, and provide a fallback if they don't provide a label. See the [amp-carousel](https://github.com/ampproject/amphtml/tree/master/extensions/amp-carousel/0.1) code to see what their fallback is, and we should just follow that pattern.

I suggest we name this label: `data-ad-label`, or whatever would be most consistent with IMA Sdk. Whenever we implement the i18n in this PR, or another, just provide a short paragraph with some links on the naming decision and we can change it if needed.

Thanks! 😄

curseagain commented:
@torch2424, I've committed the change to accept `data-ad-label` attribute as a pattern, similar to the way it's done in `amp-carousel`.  I've also updated the description in the PR so you can see the ad controls in different contexts.

Potential problem at mobile sizes where ad controls and the skip button are very close to each other?  Is it fine as-is?  Or do we need to reduce the height and font size of the controls a bit more?

Thanks for your review and feedback!!

curseagain commented:
Also, let's do "show controls on hover" in a follow-up that you can assign to me, if that's okay?

torch2424 commented:
@curseagain Thank you very much for making that change it is very much appreciated! 😄 

Thanks for adding Gifs, the previews definitely help understand what's going on 😄 

On mobile that does look a bit odd, if possible, could we reduce the size of the height of the controls a little bit more?

And updating docs and "show controls on hover" sounds good to be in a follow up PR to me. Is it alright to finish up this feature in another PR @aghassemi ?

Doing a review now, thanks again for your work! 😄

torch2424 commented on ads/google/imaVideo.js:
```diff
@@ -248,17 +257,38 @@ export function imaVideo(global, data) {
     'height': '100px',
+    'background-color': 'rgba(7, 20, 30, .7)',
+    'background':
+      'linear-gradient(0, rgba(7, 20, 30, .7) 0%, rgba(7, 20, 30, 0) 100%)',
```
Ahhh I like this approach, thank you! 😄

torch2424 commented on examples/ima-video.amp.html:
```diff
@@ -96,5 +96,28 @@ <h3>Video3 Actions</h3>
+
+    <h2>Video4</h2>
+    <div class="description">
+      Uses the Single Skippable Inline IMA Sample Tag. Also, this video with autoplay
```
Great Example, thank you!

torch2424 commented on ads/google/imaVideo.js:
```diff
@@ -210,6 +218,7 @@ export function imaVideo(global, data) {
 
   videoWidth = global./*OK*/innerWidth;
   videoHeight = global./*OK*/innerHeight;
+  adLabel = data.adLabel || 'Ad (%s of %s)';
```
Let's make sure to document later on that this will be the default 👍

torch2424 commented on ads/google/imaVideo.js:
```diff
@@ -248,17 +257,38 @@ export function imaVideo(global, data) {
+    'box-sizing': 'border-box',
+    'display': 'none',
+    'flex-grow': '1',
+    'font-family': 'Helvetica, Arial, Sans-serif',
```
Do we need this?

curseagain replied:
This style was being set in `timeDiv` so I also set it in `countdownWrapperDiv` for consistency.  However, now I've moved this style to their shared parent, `controlsDiv`, and have verified that both `timeDiv` and `countdownWrapperDiv` inherit the font-family style.

torch2424 replied:
Thank you! 😄

torch2424 commented on ads/google/imaVideo.js:
```diff
@@ -769,6 +801,24 @@ export function onAdError() {
+    remainingSeconds = '0' + remainingSeconds;
+  }
+  const label = adLabel.replace('%s', adPosition).replace('%s', totalAds);
+  countdownDiv./*OK*/innerHTML
```
@zhouyx Should it be OK to set innerHTML here?

zhouyx replied:
I am fine with using innerHTML on created div. but isn't `textContent` a better option here?

aghassemi commented:
Sorry for late reply: "show controls on hover" as a feature in a later PR would be great.

curseagain commented:
@torch2424 I've made the changes that were requested, described them in the Pull Request Description, and added some gifs.  Thanks!

torch2424 commented:
Opened up a tracking issue for the remaining work for the newly introduced features: https://github.com/ampproject/amphtml/issues/19393

cc @curseagain
